Implement proposal history api

ivanng-oursky opened the pull request:
Note: Will conflict with oursky/likedao#186 on proposal vote model

~~depends on oursky/likedao#181~~

### Changes
- reimplemented proposal status and address filters
- implemented turnout rate field for each proposal
- implemented vote distribution of address by vote option API

### Example queries

```gql
query Proposals($address: String!) {
  proposals(
    input: {first: 12, after: 0, status: Passed, address: {address: $address, isVoter: true, isDepositor: false, isSubmitter: false}}
  ) {
    edges {
      cursor
      node {
        id
        turnout
        myVote {
          option
        }
        myDeposit {
          amount {
            denom
            amount
          }
        }
      }
    }
  }
}
```

```gql
query VoteDistribution($address: String!){
  proposalVotesDistribution(address: $address) {
    yes
    no
    abstain
    noWithVeto
  }
}
```

refs oursky/likedao#176

hochiw commented on graphql-server/pkg/queries/proposal.go:
```diff
@@ -25,30 +25,64 @@ type ProposalQuery struct {
+			Column("proposal_id").
+			Where("voter_address = ?", q.scopedAddressFilter.Address)
+
+		isFirstAddressFilter := true
```
what does FirstAddress mean?

ivanng-oursky replied:
`isFirstAddressFilter` check if the address filter (ie. isDepositor, isVoter or isSubmitter) being added to the query object is the first filter being added. If it is it needs to be a `Where` rather then `WhereOr`.

hochiw replied:
it looks like they are interchangeable based on the implementation 🤔 
https://github.com/uptrace/bun/blob/e816b1a7bf524f0b9cd25579cf86e0e35bb53a02/query_select.go#L131-L139

I know they use `.Where()` + `.WhereOr()` as example in their official docs but does using only `WhereOr` produce a wrong query? bun might be unfriendly in some cases but not this unfriendly :sosad:

ivanng-oursky replied:
I am not sure how the implementation would work if only `WhereOr` is used? 

If only use `WhereOr`, in the cases where both status and address filters are applied, it would result in SQL as below, making the 2 filters applied in union, rather than their intersection.

```sql
SELECT * 
WHERE 
    status = [desired status] OR
    voter_address = [desired address] OR depositor_address = [desired address]
```

but we want 

```sql
SELECT * 
WHERE 
    status = [desired status] AND
    voter_address = [desired address] OR depositor_address = [desired address]
```

chihimng replied:
https://bun.uptrace.dev/guide/query-where.html#grouping

does this help you?

hochiw commented on graphql-server/pkg/resolvers/proposal.go:
```diff
@@ -66,6 +67,33 @@ func (r *proposalResolver) Reactions(ctx context.Context, obj *models.Proposal)
 
+func (r *proposalResolver) Turnout(ctx context.Context, obj *models.Proposal) (*string, error) {
+	stakingPool, err := pkgContext.GetQueriesFromCtx(ctx).Proposal.QueryProposalStakingPool(obj.ID)
+	if stakingPool == nil {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
```
can you see see if `err != nil` implies `stakingPool == nil`?

hochiw commented on graphql-server/pkg/models/proposal.go:
```diff
@@ -104,3 +104,12 @@ type ProposalTallyResult struct {
 	Height     int64       `bun:"column:height,notnull"`
 }
+
+type ProposalStakingPool struct {
+	bun.BaseModel `bun:"table:proposal_staking_pool_snapshot"`
+
+	ProposalID      int         `bun:"column:proposal_id,pk"`
+	BondedTokens    *bunbig.Int `bun:"column:bonded_tokens"`
+	NotBondedTokens *bunbig.Int `bun:"column:not_bonded_tokens"`
+	Height          int64       `bun:"column:height"`
+}
```
is it an existing bdjuno table or just a type? if just a custom type no need to add `bun.BaseModel` and the `bun` tags

ivanng-oursky replied:
it is a existing bdjuno table

hochiw commented on graphql-server/pkg/dataloaders/proposal.go:
```diff
@@ -22,9 +24,36 @@ type ProposalTallyResultDataloader interface {
+	proposalDepositLoader     ProposalDepositDataloader
+}
+
+// check if element is contained in array
+func contains[T comparable](arr []T, element T) bool {
+	for _, s := range arr {
+		if s == element {
+			return true
+		}
+	}
+	return false
 }
```
should be in utils

ivanng-oursky replied:
as discussed offline, will use https://pkg.go.dev/golang.org/x/exp@v0.0.0-20220613132600-b0d781184e0d/slices#Contains instead

hochiw commented on graphql-schema/proposal.graphql:
```diff
@@ -108,7 +108,15 @@ input QueryProposalsInput {
   searchTerm: String
 }
 
+type ProposalVotesDistribution {
+  yes: Float!
+  no: Float!
+  abstain: Float!
+  noWithVeto: Float!
+}
```
looks similar to `ProposalTallyResult`? maybe move `outstandingOption` to Proposal and reuse tallyResult type?

ivanng-oursky replied:
`ProposalVotesDistribution` refers to distribution of vote counts by a single wallet address across all proposals voted by this wallet address, so have a slightly different type than `ProposalTallyResult`, which refers to the vote count of all voters on a single proposal.

hochiw commented on graphql-server/pkg/queries/proposal.go:
```diff
@@ -247,6 +248,24 @@ func (q *ProposalQuery) QueryProposalVotes(proposalIDs []int, addresses []string
+		Model((*models.ProposalVote)(nil)).
+		Where("voter_address = ?", address)
+
+	count, err := query.Count(q.ctx)
+	if err != nil {
+		return nil, err
+	}
+	votes := make([]*models.ProposalVote, count)
+
+	err = query.Scan(q.ctx, &votes)
```
does query Scan work after Count? thought it would consume the query

ivanng-oursky replied:
I think it works, the query is able to return the correct vote counts when I do manual testing

chihimng replied:
https://bun.uptrace.dev/guide/query-select.html#count-rows

this exists, and why do you need to count if you won't need it in result? if you want to catch not found just check the err type and handle it

ivanng-oursky commented:
@hochiw I made a typo when summing the votes for vote distribution of a single wallet, fixed-up and rebased 🙇

hochiw commented on graphql-server/pkg/queries/proposal.go:
```diff
@@ -25,30 +25,64 @@ type ProposalQuery struct {
+				query = query.WhereOr("proposal.proposer_address = ?", q.scopedAddressFilter.Address)
+			}
+		}
+		if isFirstAddressFilter {
+			// none of the is* field are marked as true
+			query = query.Where("id IN (null)")
+		}
```
looks weird, the idea of having a particular filter configuration that only returns nothing sounds a bit off to me :sosad:

@chihimng @cychiuae any opinion from db high hands?

chihimng replied:
what was the original purpose?

ivanng-oursky replied:
To achieve the behavior where
- no address filter applied -> return unfiltered results
- address filter applied but all is* fields are false -> then no results returned

ivanng-oursky replied:
as discussed offline with @chihimng, this will be handled at the resolver level

chihimng commented on graphql-server/pkg/resolvers/proposal.go:
```diff
@@ -183,6 +243,51 @@ func (r *queryResolver) ProposalByID(ctx context.Context, id models.NodeID) (*mo
+	abstainCount := float64(0)
+	noWithVetoCount := float64(0)
+
+	for _, vote := range votes {
```
easier to just query this on sql level?

chihimng commented on graphql-schema/proposal.graphql:
```diff
@@ -93,7 +108,15 @@ input QueryProposalsInput {
   searchTerm: String
 }
 
+type ProposalVotesDistribution {
```
i think it will make the api more useful if we return actual votes count here, and let ui layer compute the ratios

chihimng commented on graphql-server/pkg/dataloaders/proposal.go:
```diff
@@ -60,9 +80,89 @@ func NewProposalDataloader(proposalQuery queries.IProposalQuery) ProposalDataloa
+			proposalIDs := []int{}
+			addresses := []string{}
+			for _, key := range keys {
+				if !slices.Contains(proposalIDs, key.ProposalID) {
```
this is slow, can you tell me the big-o notation for this?

and what would be a better way to do this?

ivanng-oursky replied:
> this is slow, can you tell me the big-o notation for this?

o(n^2) 

> and what would be a better way to do this?

as discussed offline, will use a map here instead

chihimng replied:
in particular, we can use `map[T]struct{}` in go as a set of T (since there is no native set type in go)

ivanng-oursky commented:
@hochiw updated 🙇

chihimng commented on graphql-schema/proposal.graphql:
```diff
@@ -162,7 +165,15 @@ input QueryProposalDepositsInput {
   order: ProposalDepositSort!
 }
 
+type ProposalVotesDistribution {
```
is this type still used?

chihimng commented on graphql-server/pkg/resolvers/proposal.go:
```diff
@@ -88,6 +90,61 @@ func (r *proposalResolver) MyReaction(ctx context.Context, obj *models.Proposal)
+}
+
+func (r *proposalResolver) MyDeposit(ctx context.Context, obj *models.Proposal) (*models.ProposalDeposit, error) {
+	address, err := utils.GetOperationArgumentValue(ctx, "input.address.address")
```
what if we introduce another query that returns Proposal type in the future? Assuming the parent operation here makes this un-reusable in the future

can we get the user address from request header / cookie?

either: a) require all user to auth on wallet connection, or b) send unauthenticated user address for query use in header?

chihimng commented on graphql-server/pkg/queries/proposal.go:
```diff
@@ -275,19 +298,113 @@ func (q *ProposalQuery) QueryProposalDepositTotal(id int, denom string) (bunbig.
+		case models.ProposalVoteOptionNoWithVeto:
+			distribution.NoWithVeto = voteCount.Count
+		default:
+			return nil, errors.New("invalid vote option encountered")
```
use specific error type e.g. InternalError?

elise-ng reviewed (changes requested):
please also fix all `errors.New` usages, should use typed errors

chihimng commented on graphql-server/pkg/utils/graphql.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+// get argument by name from root field (ie query or mutation) from resolver context
+func GetOperationArgument(ctx context.Context, name string) (*ast.Argument, error) {
```
I assume we can drop this file if we move on to reading unauthed user from header, or enforce all user to auth at connect wallet

chihimng commented on graphql-server/pkg/dataloaders/proposal.go:
```diff
@@ -60,9 +79,101 @@ func NewProposalDataloader(proposalQuery queries.IProposalQuery) ProposalDataloa
+				}
+			}
+
+			deposits, err := proposalQuery.QueryProposalDeposits(proposalIDs, addresses)
```
this is querying proposal ids x addresses, which could return a lot of unused results, can we make the query more precise by filtering by tuples, i.e. sql `WHERE (proposal_id, address) IN (($id1, $addr1), ($id2, $addr2))`??

ivanng-oursky committed:
graphql-server: implement proposal filter by address
refs oursky/likedao#176

ivanng-oursky committed:
graphql-server: implement proposal turnout api
refs oursky/likedao#176

ivanng-oursky committed:
graphql-server: impl proposal deposit & vote by address

- get a single proposal vote by wallet address
- get a single proposal deposit by wallet address

refs oursky/likedao#176

ivanng-oursky commented:
@chihimng @hochiw updated the following 🙇 :

- turnout calculation is now done in db level
  - added dataloader for proposal turnout
- match votes and deposits by (proposal_id, address) tuple
- use typed error for BadUserInput
- changed `myVote` and `myDeposit` fields in Proposal to `voteByAddress` and `depositByAddress`, which takes a address param

ivanng-oursky committed:
graphql-server: impl proposal vote distribution api

refs oursky/likedao#176

chihimng commented:
should follow up https://github.com/oursky/likedao/issues/229 later
